src/chk/chk_leader.c

```diff
@@ -333,20 +335,1036 @@ chk_leader_find_slowest(struct chk_instance *ins)
+
+		clue = chk_leader_locate_pool_clue(cpr);
+		ranks->rl_ranks[0] = clue->pc_rank;
+	}
+
+	rc = ds_rsvc_dist_start(DS_RSVC_CLASS_POOL, &psid, cpr->cpr_uuid, ranks,
+				cpr->cpr_healthy ? DS_RSVC_START : DS_RSVC_DICTATE,
+				false /* bootstrap */, 0 /* size */);
+
+out:
+	d_rank_list_free(ranks);
+	if (rc != 0) {
+		D_ERROR(DF_LEADER" failed to start pool service (%s) for "DF_UUIDF" at replica %d, "
+			"skip it: "DF_RC"\n",
+			DP_LEADER(ins), cpr->cpr_healthy ? "healthy" : "unhealthy",
+			DP_UUID(cpr->cpr_uuid), cpr->cpr_advice, DP_RC(rc));
+
+		cpr->cpr_skip = 1;
+		if (!cpr->cpr_healthy)
+			cbk->cb_statistics.cs_failed++;
+		chk_leader_post_repair(ins, cpr->cpr_uuid, &rc, !cpr->cpr_healthy, true);
+	} else if (!cpr->cpr_healthy) {
+		cbk->cb_statistics.cs_repaired++;
+		chk_leader_post_repair(ins, cpr->cpr_uuid, &rc, true, false);
+	}
+
+	return rc;
+}
+
+static int
+chk_leader_handle_pools_list(struct chk_sched_args *csa)
+{
+	struct chk_property		*prop = &csa->csa_ins->ci_prop;
+	struct chk_list_pool		*clp = NULL;
+	struct chk_pool_rec		*cpr;
+	struct chk_pool_rec		*tmp;
+	d_iov_t				 riov;
+	d_iov_t				 kiov;
+	int				 clp_nr;
+	int				 rc = 0;
+	int				 i;
+
+	clp_nr = ds_chk_listpool_upcall(&clp);
+	if (clp_nr < 0) {
+		rc = clp_nr;
+		clp_nr = 0;
+		goto out;
+	}
+
+	/* Firstly, handle dangling pool(s) based on the comparison between engines and MS. */
+	for (i = 0; i < clp_nr; i++) {
+		d_iov_set(&riov, NULL, 0);
+		d_iov_set(&kiov, &clp[i].clp_uuid, sizeof(uuid_t));
+		rc = dbtree_lookup(csa->csa_hdl, &kiov, &riov);
+		if (rc == 0) {
+			cpr = (struct chk_pool_rec *)riov.iov_buf;
+			/*
+			 * As for whether pool service replicas and pool label match MS or not,
+			 * they will be handled in subsequent pass.
+			 */
+			cpr->cpr_exist_on_ms = 1;
+			continue;
+		}
+
+		if (rc == -DER_NONEXIST) {
+			rc = chk_leader_dangling_pool(csa->csa_ins, clp[i].clp_uuid);
+			if (rc != 0)
+				goto out;
+		}
+
+		D_ERROR("Failed to verify pool "DF_UUIDF" existence with %s: "DF_RC"\n",
+			DP_UUID(clp[i].clp_uuid), (prop->cp_flags &
+			CHK__CHECK_FLAG__CF_FAILOUT) ? "failout" : "continue", DP_RC(rc));
+
+		if (prop->cp_flags & CHK__CHECK_FLAG__CF_FAILOUT)
+			goto out;
+	}
+
+	rc = 0;
+
+	d_list_for_each_entry_safe(cpr, tmp, &csa->csa_list, cpr_link) {
+		rc = chk_leader_handle_pool_clues(cpr);
+		if (rc != 0)
+			goto out;
+
+		if (!cpr->cpr_exist_on_ms && !cpr->cpr_skip) {
+			rc = chk_leader_orphan_pool(cpr);
+			if (rc != 0)
+				goto out;
+		}
+	}
+
+out:
+	ds_chk_free_pool_list(clp, clp_nr);
+
+	if (rc != 0)
+		D_ERROR(DF_LEADER" failed to handle pools list: "DF_RC"\n",
+			DP_LEADER(csa->csa_ins), DP_RC(rc));
+
+	return rc;
+}
+
+static int
+chk_leader_handle_pools_svc(struct chk_sched_args *csa)
+{
+	struct chk_pool_rec	*cpr;
+	struct chk_pool_rec	*tmp;
+	int			 rc = 0;
+
+	d_list_for_each_entry_safe(cpr, tmp, &csa->csa_list, cpr_link) {
+		if (!cpr->cpr_skip) {
+			rc = chk_leader_start_pool_svc(cpr);
+			if (rc != 0)
+				break;
```

Comment:
same here

Reply:
chk_leader_start_pool_svc() has already decided the next action based on the start option (failout or continue) when hit failure.